File: go/977.go
func sortedSquares(nums []int) []int {
    var flag int = -1
    for i := 0;i < len(nums) && nums[i] < 0;i ++ {
        flag = i
    }
    ans := make([]int,0,len(nums))
    for i, j := flag,flag+1; i >= 0 || j < len(nums); {
        if i < 0 {
            ans = append(ans,nums[j] * nums[j])
            j ++
        } else if j == len(nums) {
            ans = append(ans,nums[i] * nums[i])
            i --
        } else if nums[i]*nums[i] < nums[j]*nums[j] {
            ans = append(ans,nums[i] * nums[i])
            i --
        } else {
            ans = append(ans,nums[j] * nums[j])
            j ++
        }
    } 
    return ans
}

File: Cpp/977.cpp
#include <vector>
#include <algorithm>

using namespace std;


// 可以直接排序的方式
// 有点拉胯，没有运用到题目给的另一个条件
class Solution1 {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> temp1;
        vector<int> temp2;
        vector<int> res;
        int i,j;
        for(i = 0;i < nums.size();i ++) {
            if(nums[i] < 0) {
                temp1.push_back(nums[i]*nums[i]);
            } else {
                temp2.push_back(nums[i]*nums[i]);
            }
        }
        sort(temp1.begin(),temp1.end());
        i = 0;
        j = 0;
        while(i < temp1.size() && j < temp2.size()) {
            if(temp1[i] < temp2[j]) {
                res.push_back(temp1[i]);
                i ++;
            } else {
                res.push_back(temp2[j]);
                j ++;
            }
        }
        while(i < temp1.size()) {
            res.push_back(temp1[i]);
            i ++;
        }
        while(j < temp2.size()) {
            res.push_back(temp2[j]);
            j ++;
        }
        return res;
    }
};